components/browser/awesomebar/src/main/java/mozilla/components/browser/awesomebar/SuggestionsAdapter.kt

```diff
@@ -95,8 +99,13 @@ internal class SuggestionsAdapter(
         return ViewHolderWrapper(layout.createViewHolder(awesomeBar, view, viewType), view)
     }
 
+    override fun getItemId(position: Int): Long {
+        val suggestion = suggestions[position]
+        return suggestion.id.hashCode().toLong()
```

Comment:
Isn't using `hashCode` problematic? It will guarantee **stable** ids. But it doesn't guarantee you **unique** ids. Two suggestions *could* have the same hashcode? And that would lead to weird bugs when updating the adapter?

Reply:
You're correct. Currently this depends on the `id`s coming out of the providers to actually be uniue. They're prefixed with a provider identifier, and the actual ID is a URL usually, so they _should_ be unique at this point.

I agree that the current setup is a little brittle and leaks implementation details across components. It's also the "easy thing to do" here. An alternative would be to enforce uniqueness higher up the stack, closer to where it actually starts to matter. But at the end of the day, we're still depending on providers to give us good IDs to work with!

I think a reasonable thing to do here would be to document requirements around suggestion IDs clearly, static that they must include a provider identifier along with a unique-per-suggestion string (e.g. a url).

Reply:
Even with unique ids coming out of the providers we can end up with hashcode collisions, no? At the end of the day you are "mapping"/hashing Strings with "indefinite" length to a 64 bit Long.
https://android.googlesource.com/platform/libcore/+/refs/heads/master/ojluni/src/main/java/java/lang/String.java#1449

Given that hashCode aims for a uniformly random distribution of values it's probably unlikely but not impossible to have collisions?